src/lib.rs

```diff
@@ -1,40 +1,97 @@
+                )
+            })
+            .unwrap()
+    };
+}
+
+/// The basic GeoTIFF struct. This includes any metadata as well as the actual image data.
 ///
-/// It is primarily used within a routing application that needs to parse digital elevation models.
-/// As such, other use cases are NOT tested (for now).
-impl TIFF {
-    /// Opens a `.tiff` file at the location indicated by `filename`.
-    pub fn open(filename: &str) -> Result<Box<TIFF>> {
-        let tiff_reader = TIFFReader;
-        tiff_reader.load(filename)
-    }
+/// The image data has a size of width * length * num_samples
+#[derive(Debug)]
+pub struct GeoTIFF {
+    pub dimensions: (usize, usize),
+    pub num_samples: usize,
+    image_data: ImageData,
+}
+
+impl GeoTIFF {
+    pub fn read<R: Read + Seek>(reader: R) -> TiffResult<Self> {
+        let mut decoder = Decoder::new(reader)?;
 
-    /// Gets the value at a given coordinate (in pixels).
-    pub fn get_value_at(&self, lon: usize, lat: usize) -> usize {
-        self.image_data[lon][lat][0]
+        let dimensions = decoder
+            .dimensions()
+            .map(|(width, height)| (width as usize, height as usize))?;
+        let num_samples = match decoder.find_tag(Tag::SamplesPerPixel)? {
+            None => 1,
+            Some(value) => value.into_u16()? as usize,
+        };
+        let image_data = match decoder.read_image()? {
+            DecodingResult::U8(data) => ImageData::U8(data),
+            DecodingResult::U16(data) => ImageData::U16(data),
+            DecodingResult::U32(data) => ImageData::U32(data),
+            DecodingResult::U64(data) => ImageData::U64(data),
+            DecodingResult::F32(data) => ImageData::F32(data),
+            DecodingResult::F64(data) => ImageData::F64(data),
+            DecodingResult::I8(data) => ImageData::I8(data),
+            DecodingResult::I16(data) => ImageData::I16(data),
+            DecodingResult::I32(data) => ImageData::I32(data),
+            DecodingResult::I64(data) => ImageData::I64(data),
+        };
+
+        Ok(Self {
+            dimensions,
+            num_samples,
+            image_data,
+        })
     }
-}
 
-/// Overwrite default display function.
-impl fmt::Display for TIFF {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "TIFF(Image size: [{}, {}, {}], Tag data: {:?})",
-               self.image_data.len(), self.image_data[0].len(),
-               self.image_data[0][0].len(), self.ifds)
+    pub fn get_value_at<T: FromPrimitive + 'static>(&self, x: usize, y: usize, sample: usize) -> T {
+        let GeoTIFF {
+            dimensions: (image_width, ..),
+            num_samples,
+            image_data,
+            ..
+        } = self;
+
+        if sample >= *num_samples {
+            panic!(
+                "sample out of bounds: the number of samples is {} but the sample is {}",
+                num_samples, sample
+            )
+        }
```

Comment:
Could we raise an error (e.g. [`TiffError::LimitsExceeded`](https://docs.rs/tiff/0.9.1/tiff/enum.TiffError.html#variant.LimitsExceeded)) here instead of panicking?

Reply:
I thought this was similar to accessing an array element, which also panics if the index is out of bounds. Also, returning a result might make it more cumbersome to use as a library consumer. What should be checked in addition is whether `x` and `y` are less than `width` and `height`, respectively. However, this will change again if we integrate coordinate transformations.

If we decide on returning a result from this function, I'd vote for having a separate `GeoTiffError` type. As far as I understand, the `TiffError` deals with errors related to encoding/decoding, whereas here, we are handling a potential programming error.

Reply:
> I thought this was similar to accessing an array element, which also panics if the index is out of bounds.

A panic can happen if you access using `data[index]`, but not if you use `data.get(index)` which returns an `Option` (see https://doc.rust-lang.org/std/vec/struct.Vec.html#indexing and https://doc.rust-lang.org/std/primitive.slice.html#method.get.

> What should be checked in addition is whether `x` and `y` are less than `width` and `height`, respectively. However, this will change again if we integrate coordinate transformations.

Yes, ok with leaving `x > width` and `y > height` unchecked for now, and do it after coordinate transforms are implemented.

> If we decide on returning a result from this function, I'd vote for having a separate `GeoTiffError` type. As far as I understand, the `TiffError` deals with errors related to encoding/decoding, whereas here, we are handling a potential programming error.

Don't want us to get sidetracked with implementing `GeoTIFFError` and `GeoTIFFResult` (though I'm sure that will come at some point). For now, I'm happy to compromise on keeping the panic here since it matches previous behaviour, and we can refactor later to return a `Result` instead of panic.

Reply:
Just got distracted by the looks of `GeoTIFFError` and `GeoTIFFResult`. I think it should be written `GeoTiffError` and `GeoTiffResult`, according to the Rust naming conventions. Consequently, I believe the type in the crate should also be called `GeoTiff`.